ion/binarywriter.go

```diff
@@ -170,26 +170,32 @@ func (w *binaryWriter) writeBigInt(val *big.Int) error {
 
 // WriteFloat writes a floating-point value.
 func (w *binaryWriter) WriteFloat(val float64) error {
-	if val == 0 {
+	if val == 0 && !math.Signbit(val) {
+		// Positive zero is represented as just one byte.
 		return w.writeValue("Writer.WriteFloat", []byte{0x40})
 	}
 
+	var bs []byte
+
 	// Can this be losslessly represented as a float32?
 	if val == float64(float32(val)) {
-		bs := make([]byte, 5)
+		bs = make([]byte, 5)
 		bs[0] = 0x44
 
 		bits := math.Float32bits(float32(val))
 		binary.BigEndian.PutUint32(bs[1:], bits)
+	} else {
+		bs = make([]byte, 9)
+		bs[0] = 0x48
 
-		return w.writeValue("Writer.WriteFloat", bs)
-	}
-
-	bs := make([]byte, 9)
-	bs[0] = 0x48
+		bits := math.Float64bits(val)
+		binary.BigEndian.PutUint64(bs[1:], bits)
 
-	bits := math.Float64bits(val)
-	binary.BigEndian.PutUint64(bs[1:], bits)
+		// The above PutUint64() puts an unwanted 1 as the last byte for NaN.
+		if math.IsNaN(val) {
```

Comment:
If we handle `NaN` as a special case right after `-0`, we can avoid having to allocate a buffer and set/unset that byte.

```go
if val == 0 && !math.Signbit(val) {
    // Positive zero is represented as just one byte.
    return w.writeValue("Writer.WriteFloat", []byte{0x40})
} else if math.IsNaN(val) {
    return w.writeValue("Writer.WriteFloat", []byte{/* <bytes go here> */})
}
```
This also lets us represent a binary `NaN` in 4 bytes instead of 8.

Reply:
Addressed.